api/preprints/views.py

```diff
@@ -68,15 +70,51 @@
 from osf.metrics import PreprintDownload, PreprintView
 
 
+class PreprintOldVersionsImmutableMixin:
+    '''Override method to reject modify requests for old preprint versions (except for withdrawal)'''
+    def update(self, request, *args, **kwargs):
+        preprint = self.get_preprint(check_object_permissions=False)
+        if preprint.is_latest_version or preprint.machine_state == 'initial':
+            return super().update(request, *args, **kwargs)
+        message = 'User can not edit previous versions of a preprint'
+        sentry.log_message(message)
+        raise Conflict(detail=message)
+
+    def create(self, request, *args, **kwargs):
+        preprint = self.get_preprint(check_object_permissions=False)
+        if preprint.is_latest_version or preprint.machine_state == 'initial':
+            return super().create(request, *args, **kwargs)
+        message = 'User can not edit previous versions of a preprint'
+        sentry.log_message(message)
+        raise Conflict(detail=message)
+
+    def delete(self, request, *args, **kwargs):
+        preprint = self.get_preprint(check_object_permissions=False)
+        if preprint.is_latest_version or preprint.machine_state == 'initial':
+            return super().delete(request, *args, **kwargs)
+        message = 'User can not edit previous versions of a preprint'
+        sentry.log_message(message)
+        raise Conflict(detail=message)
+
+
 class PreprintMixin(NodeMixin):
     serializer_class = PreprintSerializer
     preprint_lookup_url_kwarg = 'preprint_id'
 
     def get_preprint(self, check_object_permissions=True, ignore_404=False):
-        qs = Preprint.objects.filter(guids___id=self.kwargs[self.preprint_lookup_url_kwarg], guids___id__isnull=False)
+        preprint_lookup_data = self.kwargs[self.preprint_lookup_url_kwarg].split('_v')
+
+        base_guid_id = preprint_lookup_data[0]
+        preprint_version = preprint_lookup_data[1] if len(preprint_lookup_data) > 1 else None
+        if preprint_version:
+            qs = Preprint.objects.filter(versioned_guids__guid___id=base_guid_id, versioned_guids__version=preprint_version)
+        else:
+            qs = Preprint.published_objects.filter(versioned_guids__guid___id=base_guid_id).order_by('-versioned_guids__version')
+
         try:
-            preprint = qs.select_for_update().get() if check_select_for_update(self.request) else qs.select_related('node').get()
+            preprint = qs.select_for_update().first() if check_select_for_update(self.request) else qs.select_related('node').first()
         except Preprint.DoesNotExist:
```

Comment:
This code from `develop` is buggy, query set returns None instead of raising `.DoesNotExist`. Replace try with None check

Reply:
Fixed in https://github.com/CenterForOpenScience/osf.io/pull/10885